quint/src/parsing/quintParserFrontend.ts

```diff
@@ -233,8 +213,79 @@ export function parsePhase2sourceResolution(
   for (const mods of sourceToModules.values()) {
     allModules = allModules.concat(mods)
   }
-  allModules.sort((m1, m2) => moduleRank.get(m2.name)! - moduleRank.get(m1.name)!)
-  return { ...mainPhase1Result, modules: allModules, sourceMap }
+  // sort the modules
+  const sortingResult = sortModules(allModules)
+
+  return {
+    ...mainPhase1Result,
+    errors: mainPhase1Result.errors.concat(sortingResult.errors),
+    modules: sortingResult.modules,
+    sourceMap,
+  }
+}
+
+/**
+ * Sort modules according to their imports, that is, importees should appear before importers.
+ * @param modules the modules to sort
+ * @return a structure that contains errors (if any were found) and the modules (sorted if no errors)
+ */
+function sortModules(modules: QuintModule[]): { errors: QuintError[]; modules: QuintModule[] } {
+  const idToModule = modules.reduce((accuMap, mod) => accuMap.set(mod.id, mod), ImmutMap<bigint, QuintModule>())
+  const nameToModule = modules.reduce((accuMap, mod) => accuMap.set(mod.name, mod), ImmutMap<string, QuintModule>())
+  if (nameToModule.size < modules.length) {
+    // sort the modules by their names
+    modules.sort((m1, m2) => m1.name.localeCompare(m2.name))
+    // find the first two adjacent modules that have the same name
+    let prev: Maybe<QuintModule> = none()
+    for (const mod of modules) {
+      if (prev.isJust() && prev.value.name === mod.name) {
+        const err: QuintError = {
+          code: 'QNT101',
+          message: `Two modules have a conflict on the same name: ${mod.name}`,
+          reference: mod.id,
+        }
+        return { errors: [err], modules }
+      }
+      prev = just(mod)
+    }
+    // this line should not be reachable, unless there is a bug above
+    throw new Error('This line should not be reached. Report a bug.')
+  }
```

Comment:
Shouldn't we do this during the same operation that forms the `namesToModule` map? (Eithir in the reduce, or moving that into this for loop).

 Advantages I see:

1. It would mean we only need to traverse the modules once, using the map lookup on the accumulator to check for duplicates at each addition
2. It would make the error detection clearer at the earliest stage at which the error can be detected

I would also suggest detecting all name clashes, rather than just the first two found. That seems more consistent with our general error reporting style.

Reply:
Thanks for the suggestion, @shonfeder. I wanted to ignore it at first, since having three loops instead of one is not really crucial here. But now I have refactored the code and it looks much easier to read :)